Activerecord ruby 2.7 warnings 6 0 stable batch 2

casperisfine opened the pull request:
Followup https://github.com/rails/rails/pull/38037

This eliminate all warnings in the AR test suite with sqlite3 adatper (besides the tzinfo ones).

@rafaelfranca @Edouard-chin @kamipo

byroot committed:
Fix some more Ruby 2.7 warnings in Active Record 6.0

kamipo commented on activemodel/lib/active_model/type/registry.rb:
```diff
@@ -9,15 +9,15 @@ def initialize
 
       def register(type_name, klass = nil, **options, &block)
-        block ||= proc { |_, *args| klass.new(*args) }
+        block ||= proc { |_, *args, **kwargs| klass.new(*args, **kwargs) }
```
We can't use `**kwargs` in this proc since the registered `klass.new` might not allow empty hash (e.g. `Integer.new`).
https://buildkite.com/rails/rails/builds/65803#d383432f-7419-4a27-91a9-6843aac234f1/934-942

How about to use `Proc#ruby2_keywords`?
That is provided to address this line.
https://bugs.ruby-lang.org/issues/16404

casperisfine replied:
Fixed.

kamipo commented on activerecord/lib/active_record/relation/delegation.rb:
```diff
@@ -60,15 +60,24 @@ def generate_method(method)
-            RUBY
+            if RUBY_VERSION >= "2.7"
+              module_eval <<-RUBY, __FILE__, __LINE__ + 1
+                def #{method}(...)
```
Do you prefer to use `...` with `RUBY_VERSION >= "2.7"` version checking?
Actually the warning in this generated method could be addressed by just one `ruby2_keywords(method) if respond_to?(:ruby2_keywords, true)`.
https://github.com/rails/rails/commit/2da55798ef4772977f25b739ab02567c8ded5c64#diff-563ce9140aa399b5ba97bb2a64b7c568R73

casperisfine replied:
> Do you prefer to use ... with RUBY_VERSION >= "2.7" version checking?

My reasoning is that `ruby2_keywords` is just a temporary thing to ease the transition, but `...` is here to stay. So I'd rather use it when possible.

I'm fine to use ruby2_keywords instead, if you think it's better.

kamipo replied:
How about `definition = RUBY_VERSION >= "2.7" ? "..." : "*args, &block"` to avoid code duplication?

casperisfine replied:
Done.

kamipo commented on activerecord/lib/active_record/migration/compatibility.rb:
```diff
@@ -101,7 +101,7 @@ def create_table(table_name, options = {})
             super(table_name, options: "ENGINE=InnoDB", **options)
           else
-            super
+            super(table_name, **options)
```
Can you pick 69712e82c83f25f8923dd74f4c4cd81b75616630 from master?

casperisfine replied:
Done

kamipo commented on activemodel/lib/active_model/type/registry.rb:
```diff
@@ -9,15 +9,16 @@ def initialize
 
-      def lookup(symbol, *args)
-        registration = find_registration(symbol, *args)
+      def lookup(symbol, *args, **kwargs)
```
Is this change fix any warning?

casperisfine replied:
Yes:

```
rails/activemodel/lib/active_model/type/registry.rb:18: warning: The last argument is used as the keyword parameter
rails/activerecord/lib/active_record/type/adapter_specific_registry.rb:18: warning: for `find_registration' defined here; maybe ** should be added to the call?
rails/activemodel/lib/active_model/type/registry.rb:21: warning: The last argument is used as the keyword parameter
rails/activerecord/lib/active_record/type/adapter_specific_registry.rb:33: warning: for `call' defined here; maybe ** should be added to the call?
rails/activemodel/lib/active_model/type/registry.rb:21: warning: The last argument is used as the keyword parameter
rails/activerecord/lib/active_record/type/adapter_specific_registry.rb:99: warning: for `call' defined here; maybe ** should be added to the call?
```

kamipo commented:
I know the base commit of 6633597 is #37262, but squash 2.7 warnings commits into the commit "Update for `Time#inspect`" is hard to understand the commit what does.

kamipo commented on activerecord/lib/active_record/migration/compatibility.rb:
```diff
@@ -97,11 +97,11 @@ def change_column(table_name, column_name, type, options = {})
             super(table_name, options: "ENGINE=InnoDB", **options)
           else
-            super
+            super(table_name, **options)
```
I think we can revert this line.

casperisfine replied:
Indeed.

kamipo commented on activemodel/lib/active_model/type/registry.rb:
```diff
@@ -9,15 +9,16 @@ def initialize
 
       def register(type_name, klass = nil, **options, &block)
-        block ||= proc { |_, *args| klass.new(*args) }
+        block ||= proc { |_, *args, **kwargs| klass.new(*args, **kwargs) }
```
> We can't use **kwargs in this proc since the registered klass.new might not allow empty hash (e.g. Integer.new).
https://buildkite.com/rails/rails/builds/65803#d383432f-7419-4a27-91a9-6843aac234f1/934-942

Still failed due to `**kwargs`.
https://buildkite.com/rails/rails/builds/65809#f86218b7-b104-40b1-bea5-2d305d78b2ed/996-1004

casperisfine replied:
Sorry, that was a squash mistake. I fixed it now.

casperisfine commented:
> but squash 2.7 warnings commits into the commit "Update for `Time#inspect`" is hard to understand the commit what does.

Sorry about that, I forgot I had a cherry-picked commit in my branch and squashed on it. 

I squashed everything in a single commit, because that was the simplest thing to do, but if you'd like me to retain authorship of these changes I can go over them and re-cherrypick them. Let me know.

kamipo commented:
Forwardported to master in 57ace94c42a2006b5efc0fba0539a5cc72d9a5db.
